stock_move_line_change_lot/models/stock_move_line.py

```diff
@@ -0,0 +1,166 @@
+# Copyright 2020 Camptocamp SA (http://www.camptocamp.com)
+# Copyright 2024 Jacques-Etienne Baudoux (BCIM) <je@bcim.be>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import _, models
+from odoo.exceptions import UserError
+from odoo.tools.float_utils import float_compare
+
+
+def is_lesser(value, other, rounding):
+    return float_compare(value, other, precision_rounding=rounding) == -1
+
+
+def is_bigger(value, other, rounding):
+    return float_compare(value, other, precision_rounding=rounding) == 1
+
+
+class StockMoveLine(models.Model):
+    _inherit = "stock.move.line"
+
+    def _change_lot_free_other_lines(self, need, location, product, lot, package):
+        self.ensure_one()
+        to_reassign_moves = self.move_id.browse()
+        freed_quantity = 0
+        rounding = product.uom_id.rounding
+        # We switch reservations with other move lines that have not
+        # yet been processed
+        other_lines = self.env["stock.move.line"].search(
+            [
+                ("lot_id", "=", lot.id),
+                ("product_id", "=", product.id),
+                ("location_id", "=", location.id),
+                ("package_id", "=", package.id),
+                ("state", "in", ("partially_available", "assigned")),
+                ("reserved_uom_qty", ">", 0),
+                ("qty_done", "=", 0),
+            ],
+            order="reserved_uom_qty desc",
+        )
+        # Favor lines from non-printed pickings.
+        other_lines.sorted(
+            lambda ml: (
+                ml.picking_id == self.picking_id or not ml.picking_id.printed,
+                -ml.reserved_uom_qty,
+            )
+        )
+        # Stop when required quantity is reached
+        for line in other_lines:
+            freed_quantity += line.reserved_qty
+            to_reassign_moves |= line.move_id
+            # if we leave the package level, it will try to reserve the same
+            # one again. This will trigger the deletion of the package level
+            line.package_level_id.move_line_ids.result_package_id = False
+            # unreserve qties of other lines
+            line.unlink()
+            if not is_lesser(freed_quantity, need, rounding):
+                # We reached the required quantity
+                break
+
+        return (freed_quantity, to_reassign_moves)
+
+    def write(self, vals):
+        if self.env.context.get("bypass_reservation_update"):
```

Comment:
@jbaudoux IMHO this override is too long. Deserves a separate function and a little bit splitting.